kernel/byterun/coq_interp.c

```diff
@@ -1303,6 +1313,30 @@ value coq_interprete
         Next;
       }
 
+      Instruct(MAKEUNIV) {
+        mlsize_t wosize = *pc++;
+        mlsize_t i;
+        value block;
+        print_instr("MAKEUNIV, tag=");
+        if (wosize == 0) {
+          accu = Atom(0); // FIXME forbidden
+        } else {
+          Coq_alloc_small(block, wosize, 0);
```

Comment:
While tinkering with Mtac2 yesterday, I have come across a function (`gen_match_from_to`) whose universe instance was of length 887. At this point, we are no longer in `alloc_small` territory (indeed, `Max_young_wosize` is only 256 words) but deep into segfault territory. So, you cannot blindly call `Coq_alloc_small`. For "large" values of `wosize`, you instead need to go the way of `caml_alloc_shr` + `caml_initialize`. (Just search for these functions in the file and it will become clear.)

In fact, there is no reason for `MAKEUNIV` to allocate and fill both blocks. You could just call `MAKEBLOCK` instead, which job it is. Sure, executing three opcodes instead of one will be a few cycles slower, but since you are calling `caml_callback2` anyway, performances are really the least of your concern. It is better to just reduce code duplication. It will also have the benefit of reducing the depth of the bytecode stack, since the blocks can now be created one after the other, instead of having to put both of their content on the stack first.

(Obviously, using `MAKEBLOCK` to make these large allocations will just move the segfault concern there. But that is something I can fix right now, independently of this pull request.)